src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,477 @@
+      log.debug(s"ChargingTimeOutTrigger for vehicle $vehicleId at $tick")
+      val chargingNetwork = chargingNetworkMap(vehicleManager)
+      chargingNetwork.lookupVehicle(vehicleId) match {
+        case Some(chargingVehicle) => handleEndCharging(tick, chargingVehicle)
+        case _                     => log.debug(s"Vehicle $vehicleId is already disconnected")
+      }
+      sender ! CompletionNotice(triggerId)
+
+    case ChargingPlugRequest(tick, vehicle, vehicleManager) =>
+      log.debug(s"ChargingPlugRequest received for vehicle $vehicle at $tick and stall ${vehicle.stall}")
+      if (vehicle.isBEV | vehicle.isPHEV) {
+        val chargingNetwork = chargingNetworkMap(vehicleManager)
+        // connecting the current vehicle
+        chargingNetwork.connectVehicle(tick, vehicle, sender) match {
+          case chargingVehicle @ ChargingVehicle(vehicle, _, _, station, _, _) if chargingVehicle.status == Waiting =>
+            log.debug(
+              s"Vehicle $vehicle is moved to waiting line at $tick in station $station, with {}/{} vehicles connected and {} in waiting line",
+              station.connectedVehicles.size,
+              station.zone.numChargers,
+              station.waitingLineVehicles.size
+            )
+            log.debug(s"connected vehicles: ${station.connectedVehicles.keys.mkString(",")}")
+            log.debug(s"waiting vehicles: ${station.waitingLineVehicles.keys.mkString(",")}")
+            sender ! WaitingInLine(tick, vehicle.id)
+          case chargingVehicle: ChargingVehicle =>
+            handleStartCharging(tick, chargingVehicle)
+        }
+      } else {
+        sender ! Failure(
+          new RuntimeException(
+            s"$vehicle is not a BEV/PHEV vehicle. Request sent by agent ${sender.path.name}"
+          )
+        )
+      }
+
+    case ChargingUnplugRequest(tick, vehicle, vehicleManager) =>
+      log.debug(s"ChargingUnplugRequest received for vehicle $vehicle from plug ${vehicle.stall} at $tick")
+      val physicalBounds = obtainPowerPhysicalBounds(tick, None)
+      val chargingNetwork = chargingNetworkMap(vehicleManager)
+      chargingNetwork.lookupVehicle(vehicle.id) match {
+        case Some(chargingVehicle) =>
+          val prevStartTime = chargingVehicle.latestChargingCycle.get.startTime
+          val startTime = Math.min(tick, prevStartTime)
+          val endTime = Math.max(tick, prevStartTime)
+          val duration = endTime - startTime
+          val (chargeDurationAtTick, energyToChargeAtTick) = dispatchEnergy(duration, chargingVehicle, physicalBounds)
+          chargingVehicle.processChargingCycle(startTime, energyToChargeAtTick, chargeDurationAtTick)
+          handleEndCharging(tick, chargingVehicle, Some(sender))
+        case _ =>
+          log.debug(s"Vehicle $vehicle is already disconnected at $tick")
+          sender ! UnhandledVehicle(tick, vehicle.id)
+      }
+
+    case Finish =>
+      chargingNetworkMap.foreach(_._2.clear())
```

Comment:
Should this do a final pass of the vehicle usage - I know we have `// if we still have a BEV/PHEV that is connected to a charging point,`, but want to see if it's worth a final pass through?

Reply:
I do a final pass here:
    ```
  // rescheduling the PlanningTimeOutTrigger
      sender ! CompletionNotice(
        triggerId,
        triggers ++ (if (nextTimeBin(timeBin) < endOfSimulationTime) {
                       Vector(ScheduleTrigger(PlanEnergyDispatchTrigger(nextTimeBin(timeBin)), self))
                     } else {
                       // if we still have a BEV/PHEV that is connected to a charging point,
                       // we assume that they will charge until the end of the simulation and throwing events accordingly
                       chargingNetworkMap
                         .flatMap(_._2.vehicles)
                         .map {
                           case (_, chargingVehicle @ ChargingVehicle(vehicle, stall, _, _, _, _)) =>
                             if (chargingVehicle.status == Connected) {
                               val (duration, energy) = dispatchEnergy(Int.MaxValue, chargingVehicle, physicalBounds)
                               chargingVehicle.processChargingCycle(timeBin, energy, duration)
                             }
                             ScheduleTrigger(
                               ChargingTimeOutTrigger(nextTimeBin(timeBin) - 1, vehicle.id, stall.managerId),
                               self
                             )
                         }
                         .toVector
                     })
      )
```